### PR TITLE
Define initial project structure

### PR DESCRIPTION
**Pull request by edilmedeiros**

### CODE REVIEW

**Review by itornaza — APPROVED** [review]
self-ACK

File: src/constants.rs [hunk 1]
```diff
@@ -1,2 +1,194 @@
...
+                                        network seed node(s))";
+
+pub const TESTNET_MSG: &str = "Use the test chain. Equivalent to -chain=test.";
+
+// Commands
+
+// == Control ==
+pub const HELP_COMMAND: &str = "help";
+pub const STOP: &str = "stop";
+pub const UPTIMR: &str = "uptime";
```
Comment (sb1752) [inline src/constants.rs:139]: typo? `UPTIME`

Comment (itornaza) [reply src/constants.rs:139]: Corrected it!

File: src/constants.rs [hunk 2]
```diff
@@ -1,2 +1,194 @@
...
+pub const TEST_MEMPOOL_ACCEPT: &str = "testmempoolaccept";
+
+// Commands messages
+
+// TODO: Adapt messages as we implenet the commands
+
+// == Control ==
+pub const HELP_COMMAND_MSG: &str = "help ( \"command\" )";
+pub const STOP_MSG: &str = "stop";
+pub const UPTIMR_MSG: &str = "uptime";
```
Comment (sb1752) [inline src/constants.rs:171]: same here if you didn't catch it already

Comment (itornaza) [reply src/constants.rs:171]: Keep popping them out!

File: src/parse.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    #[arg(long)]
+    pub url: String,
+
+    #[arg(short, long)]
+    pub user: String,
+
+    #[arg(short, long)]
+    pub pass: String,
+
+}
```
Comment (tareknaser) [inline src/parse.rs:28]: We should add subcommand field which would be an enum of all available commands (for now just `get_best_block_hash`)

**Review by tareknaser — COMMENTED** [review]
Thank you for putting this together 
I left some comments

File: src/commands.rs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+// External crates shortcuts
+use bitcoincore_rpc::{bitcoin, Auth, Client, Error, RpcApi};
+
+// Takes the rpc client as an input. We use a reference to avoid rust's
+// default move strategy in order to reuse the client elsewhere as well.
+// The function is expected to return the best block as a primitive number 
+// type (Not implemented yet) 
+pub fn get_best_block_hash(rpc: &Client) {
+    let best_block_hash = rpc.get_best_block_hash().unwrap();
```
[the remark this answers is not in the document]
Comment (itornaza) [reply src/commands.rs:20]: Thank you for pointing that out! I have just included this from the example in order to have the minimal skeleton without yet considering error handling at all. I have known rust for almost 10 days now, so I don't know yet what is the optimal approach towards error handling in rust apart from the Result and I am also not aware of the anyhow crate, but definitely have a look at it. I will just update the function definition and return types to remind us that we need to address error handling and expect that we are doing graceful error handling of all results in main or whatever module controls the output of our commands.

[the remark this answers is not in the document]
Comment (itornaza) [reply src/commands.rs:20]: I am including `pub fn get_best_block_hash(rpc: &Client) -> Result<BlockHash, Error>` as the function definition with BlockHash and Error defined by the `bitcoincore_rpc` module

File: src/constants.rs [hunk 5]
```diff
@@ -1,2 +1,194 @@
...
+pub const TESTNET_MSG: &str = "Use the test chain. Equivalent to -chain=test.";
+
+// Commands
+
+// == Control ==
+pub const HELP_COMMAND: &str = "help";
+pub const STOP: &str = "stop";
+pub const UPTIMR: &str = "uptime";
+
+// == Blockchain ==
```
[the remark this answers is not in the document]
Comment (itornaza) [reply src/constants.rs:141]: Very nice point! I defer this decision for later as we all get more experience on the crates involved. Please remind us later if we miss it, or even better document it as a comment in the issue #8!

File: src/constants.rs
[the remark this answers is not in the document]
Comment (itornaza) [reply src/constants.rs:1]: Just for having to find typos and review our output defaults in a single place. I don't know if I am missing something though!

File: src/commands.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tareknaser) [reply src/commands.rs:20]: > I have known rust for almost 10 days now, so I don't know yet what is the optimal approach towards error handling in rust

No worries. I think error handling in rust is very clear and neat. [This stackoverflow comment](https://stackoverflow.com/a/39478185) explains it briefly

> I am including pub fn get_best_block_hash(rpc: &Client) -> Result<BlockHash, Error> as the function definition

Thanks. That should be enough

File: src/constants.rs
[the remark this answers is not in the document]
Comment (tareknaser) [reply src/constants.rs:1]: For long help messages, I agree that would be helpful but I don't know about the commands themselves
but feel free to ignore this comment. It's a nitpick eitherway

File: src/parse.rs (hunk 3 above)
Comment (tareknaser) [reply src/parse.rs:28]: Sure!
Here is what I had in mind
```rust
#[derive(Parser, Debug)]
#[command(version, about, long_about = None)]
pub struct Args {
    #[arg(long)]
    pub url: String,

    #[arg(short, long)]
    pub user: String,

    #[arg(short, long)]
    pub pass: String,

    #[command(subcommand)]
    pub command: Commands,
}

#[derive(Debug, Subcommand)]
pub enum Commands {
    #[command(name = "getbestblockhash", about = "Get the best block hash")]
    GetBestBlockHash,
}
```

Comment (itornaza) [reply src/parse.rs:28]: Thanks! I have included it, it builds fine but instead `"getbestblockhash"`, I use our constant GET_BEST_BLOCK_HASH!

Comment (tareknaser) [reply src/parse.rs:28]: Great

File: src/commands.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sb1752) [reply src/commands.rs:20]: The benefit of this is that we can bubble up the `Result` all the way back up to the caller function so the user can see what went wrong (we can also edit the error messages with `.map_err` to be more user-friendly). Here's an example from a command line program I wrote:
See the `main.rs` function which takes in some arguments and runs the `run` function. It prints an error message if any: https://github.com/sb1752/command-line-rust-solutions/blob/main/grepr/src/main.rs
And see how all the functions are Result types in `lib.rs` and will return the error anywhere `?` is used.  https://github.com/sb1752/command-line-rust-solutions/blob/main/grepr/src/lib.rs

Note: I wrote this a few years ago, so it's using a much older version of the Clap crate.

[the remark this answers is not in the document]
Comment (itornaza) [reply src/commands.rs:20]: Good points! In the updated version I changed the function definition to return the Result and substituted the .wrap() with the ?

File: src/main.rs [hunk 6]
```diff
@@ -1,43 +1,51 @@
...
-
-    #[arg(short, long)]
-    user: String,
-
-    #[arg(short, long)]
-    pass: String,
-
-
-}
+fn main() {
```
Comment (diogocanut) [inline src/main.rs:22]:
```suggestion
fn main() -> Result<(), Box<dyn std::error::Error>> {
```

File: src/main.rs [hunk 7]
```diff
@@ -1,43 +1,51 @@
...
 
+    // Get the arguments from our wrapper parser
+    let args = get_args();
 
-fn main() {
-    let args = Args::parse();
+    // Get the rpc client from our rpc wrapper module
+    let rpc = rpc::rpc(&args.url, &args.user, &args.pass).unwrap();
```
Comment (diogocanut) [inline src/main.rs:28]:
```suggestion
    let rpc = rpc::rpc(&args.url, &args.user, &args.pass)?;
```

File: src/main.rs [hunk 8]
```diff
@@ -1,43 +1,51 @@
...
 
-    let rpc = Client::new(&args.url, Auth::UserPass(args.user, args.pass)).unwrap();
+    // Call our commands module! 
+    let best_block_hash = get_best_block_hash(&rpc);
```
Comment (diogocanut) [inline src/main.rs:31]:
```suggestion
    let best_block_hash = get_best_block_hash(&rpc)?;
```

File: src/main.rs [hunk 9]
```diff
@@ -1,43 +1,51 @@
...
     println!("best block height: {}", bestblockcount);
     let best_block_hash_by_height = rpc.get_block_hash(bestblockcount).unwrap();
     println!("best block hash by height: {}", best_block_hash_by_height);
     assert_eq!(best_block_hash_by_height, best_block_hash);
-
     let bitcoin_block: bitcoin::Block = rpc.get_by_id(&best_block_hash).unwrap();
     println!("best block hash by `get`: {}", bitcoin_block.header.prev_blockhash);
     let bitcoin_tx: bitcoin::Transaction = rpc.get_by_id(&bitcoin_block.txdata[0].txid()).unwrap();
     println!("tx by `get`: {}", bitcoin_tx.txid());
+    */
```
Comment (diogocanut) [inline src/main.rs:50]:
```suggestion
    Ok(())
```

**Review by diogocanut — COMMENTED** [review]
They way `main` is defined right now it will panic instead of raising the error to the user.

File: src/main.rs (hunk 9 above)
Comment (itornaza) [reply src/main.rs:50]: I would just wanna ask, if it ispossible for main to handle an error and return it to the OS?

**Comment by edilmedeiros** [comment]
Rebased and merged so we can keep moving.

**Pull request merged**